### PR TITLE
ipc vhost

### PR DESCRIPTION
**Pull request by nichtsfrei**
jira: SC-563

dd: inter process communication

To be able to talk to send data downstream to children (and their
children) and upstream (to the parent and main process) IPC is
introduced.

This first and simplest use case is to append found vhosts via
add_hostname.

The usage of ipc is simplified by storing all children of a process in a
global variable within processes. The handling, creation, freeing etc is
done via specialized functions of processes.

There are three header files:
- ipc_pipe.h contains the definitions for pipe specific implementation
- ipc.h contains the definitions for usage
- ipc_openvas.h contains the definitions specific for openvas

This is done to make it easier to either extend IPC or replace specific
solutions without having to change the clients too much.

To test this you can add:

```
if(description)
{
  script_oid("0.0.0.0.0.0.0.0.0.3");
  script_version("2019-11-10T15:30:28+0000");
  script_name("test-add-host-name");
  script_category(ACT_SCANNER);
  script_family("my test family");  
  script_tag(name:"some", value:"tag");
  script_tag(name:"last_modification", value:"2019-11-10 15:30:28 +0000 (Tue, 10 Nov 2020)");
  script_tag(name:"creation_date", value:"2015-03-27 12:00:00 +0100 (Fri, 27 Mar 2015)");
  script_tag(name:"cvss_base", value:"0.0");
  script_tag(name:"cvss_base_vector", value:"AV:N/AC:L/Au:N/C:N/I:N/A:N");
  script_tag(name:"qod_type", value:"remote_app");
  script_tag(name:"qod", value:"0");

  script_version("2021-08-19T02:25:52+0000");
  script_cve_id("CVE-0000-0000", "CVE-0000-0001");
  script_tag(name:"severity_vector", value:"CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H");
  script_tag(name:"severity_origin", value:"NVD");
  script_tag(name:"severity_date", value:"2020-08-07 19:36:00 +0000 (Fri, 07 Aug 2020)");
  script_xref(name:"Example", value:"GB-Test-1");
  script_xref(name:"URL", value:"https://www.greenbone.net");


  script_add_preference(name:"example", type:"entry", value:"a default string value");

  script_tag(name:"vuldetect", value:"Describes what this plugin is doing to detect a vulnerability.");

  script_tag(name:"summary", value:"A short description of the problem");
  script_tag(name:"insight", value:"Some detailed insights of the problem");
  script_tag(name:"impact", value:"Some detailed about what is impacted");

  script_tag(name:"affected", value:"Affected programs, operation system, ...");

  script_tag(name:"solution", value:"Solution description");
  script_tag(name:"solution_type", value:"Type of solution (e.g. mitigation, vendor fix)");
  script_tag(name:"solution_method", value:"how to solve it (e.g. debian apt upgrade)");
  script_tag(name:"qod_type", value:"package");
  exit(0);
}

function newhostname(name) {
  add_host_name(hostname: name);
  hn = get_host_name();
  log_message(data:hn);
}

newhostname(name: "test");

exit(0);

```
as well as 

```
if(description)
{
  script_oid("0.0.0.0.0.0.0.0.0.4");
  script_version("2019-11-10T15:30:28+0000");
  script_name("test-add-host-name");
  script_category(ACT_SCANNER);
  script_family("my test family");  
  script_tag(name:"some", value:"tag");
  script_tag(name:"last_modification", value:"2019-11-10 15:30:28 +0000 (Tue, 10 Nov 2020)");
  script_tag(name:"creation_date", value:"2015-03-27 12:00:00 +0100 (Fri, 27 Mar 2015)");
  script_tag(name:"cvss_base", value:"0.0");
  script_tag(name:"cvss_base_vector", value:"AV:N/AC:L/Au:N/C:N/I:N/A:N");
  script_tag(name:"qod_type", value:"remote_app");
  script_tag(name:"qod", value:"0");

  script_version("2021-08-19T02:25:52+0000");
  script_cve_id("CVE-0000-0000", "CVE-0000-0001");
  script_tag(name:"severity_vector", value:"CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H");
  script_tag(name:"severity_origin", value:"NVD");
  script_tag(name:"severity_date", value:"2020-08-07 19:36:00 +0000 (Fri, 07 Aug 2020)");
  script_xref(name:"Example", value:"GB-Test-1");
  script_xref(name:"URL", value:"https://www.greenbone.net");


  script_add_preference(name:"example", type:"entry", value:"a default string value");

  script_tag(name:"vuldetect", value:"Describes what this plugin is doing to detect a vulnerability.");

  script_tag(name:"summary", value:"A short description of the problem");
  script_tag(name:"insight", value:"Some detailed insights of the problem");
  script_tag(name:"impact", value:"Some detailed about what is impacted");

  script_tag(name:"affected", value:"Affected programs, operation system, ...");

  script_tag(name:"solution", value:"Solution description");
  script_tag(name:"solution_type", value:"Type of solution (e.g. mitigation, vendor fix)");
  script_tag(name:"solution_method", value:"how to solve it (e.g. debian apt upgrade)");
  script_tag(name:"qod_type", value:"package");
  scipt_dependencies("add_host_name.nasl");
  exit(0);
}

hn = get_host_name();
log_message(data:hn);

exit(0);

```

into your plugins dir; either start ospd-openvas when not running or refresh the ospd cache by executing `openvas -u` and then go to ospd-openvas source dir into smoketest.

Within the smoketest dir you can execute:

```
go run cmd/scans/main.go --cmd "start-finish" \
   --host "10.42.0.213" \
   --policy-path /var/lib/gvm/data-objects/gvmd/22.04/scan-configs \
   --oid "0.0.0.0.0.0.0.0.0.4,0.0.0.0.0.0.0.0.0.3" \
   -u /tmp/ospd.sock
```
the openvas log should contain:
```
sd   main:   INFO:2022-08-31 07h17.28 utc:947369: append_vhost: add vhost 'test' from 'NASL'
```
and the result should contain test as hostname for `0.0.0.0.0.0.0.0.0.3` results.

### CODE REVIEW

File: misc/ipc.c [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+      (*exec_ctx->func) (cctx, exec_ctx->func_arg);
+      if (exec_ctx->post_func != NULL)
+        (*exec_ctx->post_func) (cctx, exec_ctx->pre_arg);
+      switch (type)
+        {
+        case IPC_PIPE:
+          ipc_destroy (pctx);
+          break;
+        }
+      free (exec_ctx);
```
Comment (jjnicola) [inline misc/ipc.c:161]: This is producing a lot of `free(): invalid size`

**Review by jjnicola — CHANGES_REQUESTED** [review]

File: misc/ipc.c (hunk 1 above)
Comment (nichtsfrei) [reply misc/ipc.c:161]: oh, right because it is in the stack and not heap within proccess; going to fix it; thanks :)

File: src/processes.h [hunk 2]
```diff
@@ -28,6 +28,8 @@
 
 #include "../misc/ipc.h"
 
+#include "../misc/ipc.h"
```
Comment (jjnicola) [inline src/processes.h:31]: ups! twice

**Review by jjnicola — CHANGES_REQUESTED** [review]

File: nasl/nasl_host.c [hunk 3]
```diff
@@ -128,23 +135,24 @@ add_hostname (lex_ctxt *lexic)
...
-  host_pid = kb_item_get_int (lexic->script_infos->key, "internal/hostpid");
-  if (host_pid > 0)
-    kill (host_pid, SIGUSR2);
+  // send it to host process to extend vhosts list there
+  if (ipc_send (lexic->script_infos->ipc_context, IPC_MAIN, json, strlen (json))
+      < 0)
+    g_warning ("Unable to send %s to host process", lower);
 
 end_add_hostname:
+  ipc_data_destroy (hn);
```
Comment (jjnicola) [inline nasl/nasl_host.c:153]: Already called above.

**Review by jjnicola — DISMISSED** [review]
The new files misses the copyright header, and I would like to have the docstring in the new functions

File: misc/ipc.h [hunk 4]
```diff
@@ -0,0 +1,89 @@
+#ifndef IPC_H
+#define IPC_H
```
Comment (ArnoStiefvater) [inline misc/ipc.h:2]: As per https://github.com/greenbone/openvas-scanner/pull/1180 the new format is MISC_IPC_H. Same for all other header files.

**Review by jjnicola — APPROVED** [review]

**Pull request merged**